base/compiler/tfuncs.jl

```diff
@@ -1003,11 +1008,16 @@ function _fieldtype_tfunc(@nospecialize(s), exact::Bool, @nospecialize(name))
         return tmerge(_fieldtype_tfunc(rewrap(u.a, s), exact, name),
                       _fieldtype_tfunc(rewrap(u.b, s), exact, name))
     end
-    u isa DataType || return Type
-    u.abstract && return Type
+    u isa DataType || return Any
+    if u.abstract
+        # Abstract types have no fields
+        exact && return Bottom
+        isconstType(u) && return fieldtype_tfunc(rewrap(u.parameters[1], s), name)
```

Comment:
Hm? This one confuses me. If `fieldtype_tfunc` is called with `s0==Type{Type{Foo}}`, then we'd get here with `u == Type{Int}` (if it wasn't for `exact`being true) and return `fieldtype_tfunc(Foo, name)`? That looks quite wrong. Did I miss something? For which `s0` would we get here anyway, i.e. `isconstType(u) && !exact`? (I feel like I'm overlooking something fairly obvious...)

Reply:
This seemed very clear to me yesterday, but now I'm confused myself... Maybe my brain was just fried.

Reply:
Alright, I think @JeffBezanson and I have come up with a correct version of this.